CLI: add optional analytics tracking

samiff opened the pull request:
## Proposed changes:

- Adds optional analytics tracking in Jetpack CLI using Tracks.
- Adds `jetpack cli analytics <on|off>` to set analytics tracking preference.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

https://github.com/Automattic/garage-backlog/issues/57

## Does this pull request change what data or activity we track or use?

Yes, this PR adds optional Tracks data collection to Jetpack CLI. The purpose of this collection is to provide general metrics to inform on Jetpack CLI development projects.

## Testing instructions:

- Pull down changes and run `jetpack rsync --watch` which should prompt you to enable analytics (for testing purposes say yes). This should only ask you once.
- You can bail out of the sync process, we just wanted to fire the `tracks()` function.
- After some time (5-10 minutes), you should see a `jetpack_cli_rsync_watch` record over at the Tracks Live View page.
- The recorded event should have a `uuid` in the `eventprops` which matches the one stored in `~/.config/configstore/automattic/jetpack-cli/cli.json`
- Try running `jetpack cli analytics` which should ask you to specify the 'on' or 'off' positional. Disable for now.
- If analytics are disabled, then `jetpack rsync --watch` should not fire the Tracks event.

## Todos:

- [x] Remove 'test' prefix from event name.
- [x] After merge, formally validate and register the new Tracks event. [ Done in https://github.com/Automattic/tracks-events-registration/pull/2328 ]
- [x] After merge, give notice to devs. [ Done in pdWQjU-Ks-p2 ]
- [x] Ensure any privacy considerations have been reviewed.

samiff committed:
Fix comment typo

samiff committed:
Add jetpack cli 'analytics' subcommand

samiff committed:
Add analytics.js helper file

samiff committed:
Small refactor

samiff committed:
Adjust output wording

samiff committed:
Add axios package

samiff committed:
Add tracks function to rsync for --watch option

samiff committed:
Refactor analytics.js

- Store a uuid for the user with Configstore
- Send the Tracks POST request

samiff committed:
Add privacy mention to readme

anomiex commented on tools/cli/helpers/analytics.js:
```diff
@@ -0,0 +1,79 @@
+import crypto from 'crypto';
+import axios from 'axios';
```
Can we use native [fetch](https://nodejs.org/api/globals.html#fetch) rather than axios?

See also https://github.com/Automattic/jetpack/pull/34238#issuecomment-1822357122

samiff replied:
Great suggestion 👍 c683913475de6f0eeadb0cf38bffdeb3fdbf3bed

samiff committed:
Use node native fetch over axios

samiff committed:
Merge remote-tracking branch 'origin/trunk' into add/jp-cli-analytics

anomiex commented on tools/cli/commands/cli.js:
```diff
@@ -138,6 +157,17 @@ export function cliDefine( yargs ) {
+			.command(
+				'analytics',
+				'Toggle analytics tracking for the CLI',
+				() => {},
```
I wonder if it would be useful to have it be like `jetpack cli analytics on` and `jetpack cli analytics off` instead of just a toggle.

samiff replied:
An explicit positional is probably more straightforward 👍 dbb322d9a80035a8d8ca6017cace6a621dae686f

anomiex commented on tools/cli/helpers/analytics.js:
```diff
@@ -0,0 +1,76 @@
+ *
+ * @returns {boolean} Whether analytics tracking is enabled.
+ */
+async function checkAnalyticsEnabled() {
```
I wonder if it would be worth sticking a call to this in `tools/cli/bin/jetpack.js` to ask on the first run, instead of waiting until they're in the middle of something.

Also, probably if `process.env.CI` is set we should just assume "no" so we never risk running into the prompt in CI.

samiff replied:
> sticking a call to this in tools/cli/bin/jetpack.js to ask on the first run

That makes sense, done in da007a5f2ae976c5d26b6f4e030a3d5728cd5e50

> if process.env.CI is set we should just assume "no"

Ah good catch 👍 5d015b7b4fdfe1a453616f1f4305dab8819f7e99

anomiex commented on tools/cli/helpers/analytics.js:
```diff
@@ -0,0 +1,76 @@
+		await fetch( 'https://public-api.wordpress.com/rest/v1.1/tracks/record?http_envelope=1', {
+			method: 'POST',
+			headers: {
+				'User-Agent': 'Node.js Jetpack CLI',
```
Probably doesn't matter for anything, but if we want to be compliant with the definition for the User-Agent header the product token shouldn't contain spaces. So something like this may be better
```suggestion
				'User-Agent': 'Jetpack-CLI',
```

samiff committed:
Merge remote-tracking branch 'origin/trunk' into add/jp-cli-analytics

samiff committed:
Don't run analytics for CI

samiff committed:
Ask to enable analytics earlier on

Hopefully encountering the prompt earlier on will
avoid disrupting users in the middle of workflows.

samiff committed:
Change analytics pref to a positional not a toggle

samiff committed:
Update tools/cli/helpers/analytics.js

Co-authored-by: Brad Jorsch <anomiex@users.noreply.github.com>

samiff committed:
Remove 'test' prefix from event name

anomiex commented on tools/cli/README.md:
```diff
@@ -136,4 +136,4 @@ In addition, when you run `jetpack draft disable`, it will offer to run the pre-
 ## Privacy
 
-To effectively improve the Jetpack CLI tool, we optionally collect [general analytics data](https://jetpack.com/support/privacy/#analytics). Jetpack CLI will ask before enabling data collection, and you can always disable or enable the data collection preference by running: `jetpack cli analytics`
+To effectively improve the Jetpack CLI tool, we optionally collect [general analytics data](https://jetpack.com/support/privacy/#analytics). Jetpack CLI will ask before enabling data collection, and you can always disable or enable the data collection preference by running: `jetpack cli analytics <on|off>`
```
Total nitpick: since the command says `<on|off>`, saying "enable or disable" matches the order.
```suggestion
To effectively improve the Jetpack CLI tool, we optionally collect [general analytics data](https://jetpack.com/support/privacy/#analytics). Jetpack CLI will ask before enabling data collection, and you can always enable or disable the data collection preference by running: `jetpack cli analytics <on|off>`
```

anomiex commented on tools/cli/commands/cli.js:
```diff
@@ -101,17 +101,19 @@ function cliUnlink( options ) {
  */
-function cliAnalytics() {
-	const analyticsEnabled = ! configStore.get( 'analyticsEnabled' );
+function cliAnalytics( preference = null ) {
```
Confusing to have it default to null when it's always supposed to be a string.
```suggestion
function cliAnalytics( preference ) {
```

anomiex commented on tools/cli/bin/jetpack.js:
```diff
@@ -71,5 +72,10 @@ try {
+/**
+ * Check if the user has been asked to enable analytics tracking.
+ */
+await checkAnalyticsEnabled();
```
I note this can result in slightly silly behavior if the very first thing you run is `jetpack cli analytics on|off` 😀

We might just decide that's enough of a rare edge case that it's not worth caring about. OTOH, if we want to care, I dropped a suggestion later.

anomiex commented on tools/cli/helpers/analytics.js:
```diff
@@ -0,0 +1,80 @@
+		return false;
+	}
+
+	if ( configStore.get( 'askedToEnableAnalytics' ) === undefined ) {
```
Re the silly behavior mentioned above, if we want to care we might do like this:
```suggestion
	if ( configStore.get( 'askedToEnableAnalytics' ) === undefined ) {
		// Skip asking if they're trying to run `jetpack cli analytics` itself.
		const args = process.argv.filter( arg => ! arg.startsWith( '-' ) );
		if ( args[2] === 'cli' && args[3] === 'analytics' ) {
			return false;
		}

```

Although, to follow up on that, we should then move lines 31–37 and 40–42 into a separate function to also call from `tools/cli/commands/cli.js`, instead of partially duplicating the code there. We might want to do that anyway.

samiff replied:
Thanks, that would be a bit cleaner, done in bd0aaa7

anomiex reviewed (commented):
A few nitpicks inline, otherwise LGTM

samiff committed:
Wording

Co-authored-by: Brad Jorsch <anomiex@users.noreply.github.com>

samiff committed:
Minor code quality

Co-authored-by: Brad Jorsch <anomiex@users.noreply.github.com>

samiff committed:
Merge remote-tracking branch 'origin/trunk' into add/jp-cli-analytics

samiff committed:
Minor quality/dry refactor

anomiex reviewed (approved):
LGTM, let's do it!
